block_darwin.go

```diff
@@ -0,0 +1,250 @@
+	TotalSize                                   int64  // 499963174912
+	VolumeAllocationBlockSize                   int64  // 4096
+	VolumeName                                  string // Macintosh HD
+	VolumeSize                                  int64  // 499963174912
+	VolumeUUID                                  string // some-uuid-foo-bar
+	WholeDisk                                   bool   // false
+	Writable                                    bool   // true
+	WritableMedia                               bool   // true
+	WritableVolume                              bool   // true
+	// also has a SMARTDeviceSpecificKeysMayVaryNotGuaranteed dict with various info
+	// NOTE: VolumeUUID sometimes == DiskUUID, but not always. So far Content is always a different UUID.
+}
+
+type ioregPlist struct {
+	// there's a lot more than just this...
+	ModelNumber  string `plist:"Model Number"`
+	SerialNumber string `plist:"Serial Number"`
+	VendorName   string `plist:"Vendor Name"`
+}
+
+func (ctx *context) getDiskUtilListPlist() (diskUtilListPlist, error) {
+	out, err := exec.Command("diskutil", "list", "-plist").Output()
+	if err != nil {
+		return diskUtilListPlist{}, errors.Wrap(err, "diskutil list failed")
+	}
+
+	var data diskUtilListPlist
+	if _, err := plist.Unmarshal(out, &data); err != nil {
+		return diskUtilListPlist{}, errors.Wrap(err, "diskutil list plist unmarshal failed")
+	}
+
+	return data, nil
+}
+
+func (ctx *context) getDiskUtilInfoPlist(device string) (diskUtilInfoPlist, error) {
+	out, err := exec.Command("diskutil", "info", "-plist", device).Output()
+	if err != nil {
+		return diskUtilInfoPlist{}, errors.Wrapf(err, "diskutil info for %q failed", device)
+	}
+
+	var data diskUtilInfoPlist
+	if _, err := plist.Unmarshal(out, &data); err != nil {
+		return diskUtilInfoPlist{}, errors.Wrapf(err, "diskutil info plist unmarshal for %q failed", device)
+	}
+
+	return data, nil
+}
+
+func (ctx *context) getIoregPlist(ioDeviceTreePath string) (ioregPlist, error) {
+	elems := strings.Split(ioDeviceTreePath, "/")
+	name := elems[len(elems)-1]
+
+	args := []string{
+		"ioreg",
+		"-a",      // use XML output
+		"-d", "1", // limit device tree output depth to root node
+		"-r",       // root device tree at matched node
+		"-n", name, // match by name
+	}
+	out, err := exec.Command(args[0], args[1:]...).Output()
+	if err != nil {
+		return ioregPlist{}, errors.Wrapf(err, "ioreg query for %q failed", ioDeviceTreePath)
+	}
+
+	var data []ioregPlist
+	if _, err := plist.Unmarshal(out, &data); err != nil {
+		return ioregPlist{}, errors.Wrapf(err, "ioreg unmarshal for %q failed", ioDeviceTreePath)
+	}
+	if len(data) != 1 {
+		err := errors.Errorf("ioreg unmarshal resulted in %d I/O device tree nodes (expected 1)", len(data))
+		return ioregPlist{}, err
+	}
+
+	return data[0], nil
+}
+
+func (ctx *context) makePartition(disk, s diskOrPartitionPlistNode, isAPFS bool) (*Partition, error) {
+	if s.Size < 0 {
```

Comment:
If you made diskOrPartitionPlistNode.Size a `uint64` this wouldn't be necessary.

Reply:
Except I don't control the input, it's not open source (as far as I can tell) so I can't determine if it's possible to ever receive a negative value (I've never seen one, and can't think of why there would be, but that doesn't mean there never would be). This way makes catching the error explicit and the root cause obvious rather than a "cannot unmarshal to type" or such.

Reply:
I don't feel strongly about this though. If you still want me to change it, I will.

Reply:
@jpo-joyent yeah, you make a good point about the relative obscurity of the error that would be produced if the (closed-source) plist response ever contained a negative size. I suppose that's the suck about dealing with closed-source software :)

OK, I'm cool leaving this as-is.